### PR TITLE
Settle vault

### PR DESCRIPTION
**Pull request by haythem96**
# Task: Feature Name:

Controller: settle vault

## High Level Description

This PR add settle vault functionality.
Should be reviewed/merged after #128 .

### Code

- [x] Unit test 100% coverage
- [x] Does your code follow the naming and code documentation guidelines?

### Documentation

- [x] Is your code up to date with the spec? 
- [x] Have you added your tests to the testing doc?

### CODE REVIEW

File: contracts/Controller.sol [hunk 1]
```diff
@@ -522,13 +544,75 @@ contract Controller is ReentrancyGuard, Ownable {
...
+        address marginPoolModule = AddressBookInterface(addressBook).getMarginPool();
+        MarginPoolInterface marginPool = MarginPoolInterface(marginPoolModule);
+
+        marginPool.transferToUser(otoken.collateralAsset(), _args.receiver, payout);
+
+        emit Exercise(_args.otoken, msg.sender, _args.receiver, otoken.collateralAsset(), _args.amount, payout);
+    }
 
     /**
      * @notice settle vault option
```
Comment (antoncoding) [inline contracts/Controller.sol:567]: > 
```suggestion
     * @notice settle vault after expiry
```

File: contracts/Controller.sol [hunk 2]
```diff
@@ -548,4 +632,19 @@ contract Controller is ReentrancyGuard, Ownable {
     function isNotEmpty(address[] memory _array) internal pure returns (bool) {
```
Comment (antoncoding) [inline contracts/Controller.sol:632]:
```suggestion
    function _isNotEmpty(address[] memory _array) internal pure returns (bool) {
```

File: contracts/Controller.sol [hunk 3]
```diff
@@ -522,13 +544,75 @@ contract Controller is ReentrancyGuard, Ownable {
...
+        address marginPoolModule = AddressBookInterface(addressBook).getMarginPool();
+        MarginPoolInterface marginPool = MarginPoolInterface(marginPoolModule);
+
+        if (isNotEmpty(vault.longOtokens)) {
+            OtokenInterface longOtoken = OtokenInterface(vault.longOtokens[0]);
+
+            longOtoken.burnOtoken(marginPoolModule, vault.longAmounts[0]);
+        }
+
+        vaults[_args.owner][_args.vaultId]._clearVault();
```
Comment (antoncoding) [inline contracts/Controller.sol:603]:
```suggestion
        vaults[_args.owner][_args.vaultId]._clearVault();
        if (isNotEmpty(vault.longOtokens)) {
            OtokenInterface longOtoken = OtokenInterface(vault.longOtokens[0]);

            longOtoken.burnOtoken(marginPoolModule, vault.longAmounts[0]);
        }
```
Better update state first before external calls

File: contracts/Controller.sol [hunk 4]
```diff
@@ -522,13 +544,75 @@ contract Controller is ReentrancyGuard, Ownable {
...
+
+        emit VaultSettled(
+            address(shortOtoken),
+            _args.owner,
+            _args.to,
+            shortOtoken.collateralAsset(),
+            _args.vaultId,
+            payout
+        );
+    }
```
Comment (antoncoding) [inline contracts/Controller.sol:615]: Save some gas by avoiding duplicated external calls
```suggestion
        address collateral = shortOtoken.collateralAsset();
        marginPool.transferToUser(collateral, _args.to, payout);

        emit VaultSettled(
            address(shortOtoken),
            _args.owner,
            _args.to,
            collateral,
            _args.vaultId,
            payout
        );
    }
```

**Review by antoncoding — APPROVED** [review]
LGTM

File: contracts/Controller.sol (hunk 3 above)
Comment (haythem96) [reply contracts/Controller.sol:603]: Yeah, but if I clear vault first, then `vault.longOtokens` will be empty, also we are making a call to a long otoken in a vault that is depositing through the protocol, so all long otoken that are there are whitelisted and verified.

File: test/controller.test.ts [hunk 5]
```diff
@@ -2299,6 +2284,610 @@ contract('Controller', ([owner, accountOwner1, accountOperator1, random]) => {
...
+        'Margin pool collateral asset balance mismatch',
+      )
+      assert.equal(
+        senderBalanceAfter.minus(senderBalanceBefore).toString(),
+        payout.toString(),
+        'Sender collateral asset balance mismatch',
+      )
+    })
+  })
+
```
Comment (aparnakr) [inline test/controller.test.ts:2890]: add a test to check multiple vaults can be settled at the same time

Comment (aparnakr) [reply test/controller.test.ts:2890]: add test to check an ITM option and an OTM option

**Review by aparnakr — APPROVED** [review]
Looks good to merge. 

Add 2 more tests: 
1. Test an OTM expiry and ITM expiry 
2. Multiple settle vaults in one operate call

**Comment by haythem96** [comment]
> Looks good to merge.
> 
> Add 2 more tests:
> 
> 1. Test an OTM expiry and ITM expiry
> 2. Multiple settle vaults in one operate call

Added.

**Pull request merged**